promoted_python_delivery_client/client/client.py

```diff
@@ -0,0 +1,199 @@
+from promoted_python_delivery_client.model.timing import Timing
+from promoted_python_delivery_client.model.traffic_type import TrafficType
+
+
+SERVER_VERSION = "python.1.0.0"
+
+# Executor to run metrics logging in the background.
+DEFAULT_METRICS_THREAD_POOL_SIZE = 5
+
+# Default number of maximum request insertion passed to Delivery API.
+DEFAULT_MAX_REQUEST_INSERTIONS = 1000
+
+
+class PromotedDeliveryClient:
+    def __init__(self,
+                 delivery_endpoint: str,
+                 delivery_api_key: str,
+                 metrics_endpoint: str,
+                 metrics_api_key: str,
+                 delivery_timeout_millis: Optional[int] = None,
+                 metrics_timeout_millis: Optional[int] = None,
+                 max_request_insertions: Optional[int] = DEFAULT_MAX_REQUEST_INSERTIONS,
+                 shadow_traffic_delivery_rate: Optional[float] = 0,
+                 perform_checks: Optional[bool] = False) -> None:
+        self.metrics_endpoint = metrics_endpoint
+        self.metrics_api_key = metrics_api_key
+        self.metrics_timeout_millis = metrics_timeout_millis
+        self.max_request_insertions = max_request_insertions
+        self.sampler = Sampler()
+        self.sdk_delivery = SDKDelivery()
+        self.api_delivery = APIDelivery(delivery_endpoint, delivery_api_key, delivery_timeout_millis)
+        self.api_metrics = APIMetrics(metrics_endpoint, metrics_api_key, metrics_timeout_millis)
+
+        if shadow_traffic_delivery_rate is not None and (shadow_traffic_delivery_rate < 0 or shadow_traffic_delivery_rate > 1):
+            raise ValueError("shadow traffic delivery rate must be between 0 and 1")
+        self.shadow_traffic_delivery_rate = shadow_traffic_delivery_rate
+
+        self.perform_checks = perform_checks if perform_checks is not None else False
+
+        self.executor = concurrent.futures.ThreadPoolExecutor(DEFAULT_METRICS_THREAD_POOL_SIZE)
+
+    def deliver(self, request: DeliveryRequest) -> DeliveryResponse:
+        should_send_shadow_traffic = self._should_send_shadow_traffic()
+        exec_svr = ExecutionServer.SDK
+        cohort_membership = self._check_cohort_membership(request)
+        attempted_delivery_api = False
+        self._fill_request_fields(request.request)
+
+        if request.only_log or not self._should_apply_treatment(cohort_membership):
+            resp = self.sdk_delivery.run_delivery(request)
+        else:
+            attempted_delivery_api = True
+            try:
+                resp = self.api_delivery.run_delivery(request)
+                exec_svr = ExecutionServer.API
+            except Exception as ex:
+                logging.error("Error calling delivery", exc_info=ex)
+                resp = self.sdk_delivery.run_delivery(request)
+
+        # If delivery happened client-side, log the insertions to metrics API.
+        if exec_svr != ExecutionServer.API or cohort_membership is not None:
```

Comment:
Since exec_svr is set to API after the call to run_delivery, I believe this is correct, because if an exception happened we would have jumped around that line ~83.